# Convolutional Network with one layer

In [1]:
from __future__ import print_function
import keras
from keras.datasets import mnist
from keras.models import Sequential
from keras.layers import Dense, Dropout, Flatten
from keras.layers import Conv2D, MaxPooling2D
from keras import backend as K

/home/fernanda/anaconda3/lib/python3.6/site-packages/h5py/__init__.py:36: FutureWarning: Conversion of the second argument of issubdtype from `float` to `np.floating` is deprecated. In future, it will be treated as `np.float64 == np.dtype(float).type`.
  from ._conv import register_converters as _register_converters
Using TensorFlow backend.


## Dataset: Loading and preparation

In [2]:
# the data, split between train and test sets
(x_train, y_train), (x_test, y_test) = mnist.load_data()

# input image dimensions
img_rows, img_cols = 28, 28

if K.image_data_format() == 'channels_first':
    x_train = x_train.reshape(x_train.shape[0], 1, img_rows, img_cols)
    x_test = x_test.reshape(x_test.shape[0], 1, img_rows, img_cols)
    input_shape = (1, img_rows, img_cols)
else:
    x_train = x_train.reshape(x_train.shape[0], img_rows, img_cols, 1)
    x_test = x_test.reshape(x_test.shape[0], img_rows, img_cols, 1)
    input_shape = (img_rows, img_cols, 1)

x_train = x_train.astype('float32')
x_test = x_test.astype('float32')
x_train /= 255
x_test /= 255
print('x_train shape:', x_train.shape)
print(x_train.shape[0], 'train samples')
print(x_test.shape[0], 'test samples')
print (input_shape)

# convert class vectors to binary class matrices
num_classes = 10
y_train = keras.utils.to_categorical(y_train, num_classes)
y_test = keras.utils.to_categorical(y_test, num_classes)

batch_size = 128
epochs = 12

x_train shape: (60000, 28, 28, 1)
60000 train samples
10000 test samples
(28, 28, 1)


# Conv2D + ReLu

<img src="conv1.png">

In [7]:
model = Sequential()
# Input Layer
input_shape=input_shape
# Hidden Layer
model.add(Conv2D(32, kernel_size=(3, 3), activation='relu', input_shape=input_shape))
model.add(Flatten())
model.add(Dense(512, activation='relu'))
# Output Layer
model.add(Dense(num_classes, activation='softmax'))
model.summary()

_________________________________________________________________
Layer (type)                 Output Shape              Param #   
conv2d_1 (Conv2D)            (None, 26, 26, 32)        320       
_________________________________________________________________
flatten_1 (Flatten)          (None, 21632)             0         
_________________________________________________________________
dense_1 (Dense)              (None, 512)               11076096  
_________________________________________________________________
dense_2 (Dense)              (None, 10)                5130      
Total params: 11,081,546
Trainable params: 11,081,546
Non-trainable params: 0
_________________________________________________________________


In [8]:
#model.compile(loss='categorical_crossentropy', metrics=['accuracy'], optimizer='adam')
model.compile(loss='categorical_crossentropy', metrics=['accuracy'], optimizer='sgd')

model.fit(x_train, y_train,
          batch_size=batch_size,
          epochs=epochs,
          verbose=1,
          validation_data=(x_test, y_test))

score = model.evaluate(x_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

Train on 60000 samples, validate on 10000 samples
Epoch 1/12
60000/60000 [==============================] - 27s 446us/step - loss: 0.7277 - acc: 0.8259 - val_loss: 0.3070 - val_acc: 0.9146
Epoch 2/12
60000/60000 [==============================] - 25s 415us/step - loss: 0.2879 - acc: 0.9175 - val_loss: 0.2411 - val_acc: 0.9323
Epoch 3/12
60000/60000 [==============================] - 25s 415us/step - loss: 0.2340 - acc: 0.9336 - val_loss: 0.2116 - val_acc: 0.9391
Epoch 4/12
60000/60000 [==============================] - 25s 416us/step - loss: 0.1980 - acc: 0.9434 - val_loss: 0.1741 - val_acc: 0.9492
Epoch 5/12
60000/60000 [==============================] - 25s 416us/step - loss: 0.1707 - acc: 0.9511 - val_loss: 0.1618 - val_acc: 0.9540
Epoch 6/12
60000/60000 [==============================] - 25s 416us/step - loss: 0.1501 - acc: 0.9564 - val_loss: 0.1383 - val_acc: 0.9609
Epoch 7/12
60000/60000 [==============================] - 25s 416us/step - loss: 0.1335 - acc: 0.9615 - val_loss: 0.

# Conclusion

The mean goal convolutional neural network is extract features of input datas. This arquiteture doesn't depend on the input datas to be flattened, which could be a advantage. The weights are considered a volume in the images. For example: image = (28x28x1), kernel = (3x3), bias=1 and we would like have a 32 channels. So, we will have (3*3*32)+1*32 = 320 weights. 

# References

- http://cs231n.github.io/convolutional-networks/